File: internal/ui/stream.go
package ui

import (
	"fmt"
	"strings"
	"time"

	"github.com/charmbracelet/bubbles/spinner"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/claby2/hladmin/internal/executor"
)

// streamModel runs a command on many hosts in parallel, printing each host's full
// output block to scrollback as it finishes while keeping a live spinner block for
// the hosts still running.
type streamModel struct {
	hosts     []string
	command   string
	spinner   spinner.Model
	starts    []time.Time
	done      []bool
	results   []executor.Result
	remaining int
}

func (m streamModel) Init() tea.Cmd {
	cmds := []tea.Cmd{m.spinner.Tick}
	for i, host := range m.hosts {
		cmds = append(cmds, runHostCmd(i, host, m.command))
	}
	return tea.Batch(cmds...)
}

func (m streamModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case hostDoneMsg:
		m.done[msg.index] = true
		m.results[msg.index] = msg.result
		m.remaining--
		// Push the finished block into scrollback above the live region.
		printBlock := tea.Println(renderResultBlock(msg.result))
		if m.remaining == 0 {
			return m, tea.Sequence(printBlock, tea.Quit)
		}
		return m, printBlock
	case spinner.TickMsg:
		var cmd tea.Cmd
		m.spinner, cmd = m.spinner.Update(msg)
		return m, cmd
	case tea.KeyMsg:
		if msg.Type == tea.KeyCtrlC {
			return m, tea.Quit
		}
	}
	return m, nil
}

func (m streamModel) View() string {
	var b strings.Builder
	for i, host := range m.hosts {
		if m.done[i] {
			continue
		}
		b.WriteString(fmt.Sprintf("%s %s  %s\n",
			m.spinner.View(),
			Hostname.Render(host),
			Secondary.Render(FormatDuration(time.Since(m.starts[i])))))
	}
	return b.String()
}

// RunStreaming executes command on hosts in parallel. On a TTY it shows a live
// spinner block for running hosts and streams each host's output block as it
// completes. On a non-TTY it prints each block as it finishes with no animation.
func RunStreaming(hosts []string, command string) ([]executor.Result, error) {
	if err := executor.VerifyHostsAndCommand(hosts, command); err != nil {
		return nil, err
	}

	if !IsTerminal() {
		return runStreamingPlain(hosts, command), nil
	}

	now := time.Now()
	starts := make([]time.Time, len(hosts))
	for i := range starts {
		starts[i] = now
	}

	m := streamModel{
		hosts:     hosts,
		command:   command,
		spinner:   newSpinner(),
		starts:    starts,
		done:      make([]bool, len(hosts)),
		results:   make([]executor.Result, len(hosts)),
		remaining: len(hosts),
	}

	final, err := tea.NewProgram(m).Run()
	if err != nil {
		return nil, err
	}
	return final.(streamModel).results, nil
}

func runStreamingPlain(hosts []string, command string) []executor.Result {
	results := make([]executor.Result, len(hosts))
	done := make(chan int, len(hosts))
	for i, host := range hosts {
		go func(i int, host string) {
			results[i] = executor.RunOnHost(host, command)
			done <- i
		}(i, host)
	}
	for range hosts {
		i := <-done
		fmt.Println(renderResultBlock(results[i]))
		fmt.Println()
	}
	return results
}

File: src/ui/stream.rs
use crate::executor::{self, ExecResult};
use crate::ui::render::{format_duration, render_result_block};
use crate::ui::styles;
use anyhow::Result;
use indicatif::{MultiProgress, ProgressBar, ProgressDrawTarget, ProgressStyle};
use std::sync::mpsc::RecvTimeoutError;
use std::time::{Duration, Instant};

/// Spinner cadence shared by the streaming and live-table views.
pub const TICK_INTERVAL: Duration = Duration::from_millis(100);

/// Executes command on hosts in parallel. On a TTY it shows a live spinner block
/// for running hosts and streams each host's output block into scrollback as it
/// completes. On a non-TTY it prints each block as it finishes with no animation.
pub fn run_streaming(hosts: &[String], command: &str) -> Result<Vec<ExecResult>> {
    if !styles::is_terminal() {
        return run_streaming_plain(hosts, command);
    }

    let start = Instant::now();
    let rx = executor::run_parallel(hosts, command)?;

    let mp = MultiProgress::with_draw_target(ProgressDrawTarget::stdout());
    // new_spinner comes preconfigured with indicatif's default spinner style,
    // whose template is exactly the "{spinner} {msg}" layout used here.
    let bars: Vec<ProgressBar> = hosts
        .iter()
        .map(|host| {
            let pb = mp.add(ProgressBar::new_spinner());
            pb.set_message(running_message(host, start));
            pb
        })
        .collect();

    // A single-host run needs no tally; for multiple hosts a summary line
    // below the spinners keeps overall progress visible even once completed
    // blocks start filling the scrollback.
    let summary = (hosts.len() > 1).then(|| {
        let pb = mp.add(ProgressBar::new_spinner());
        pb.set_style(ProgressStyle::with_template("{msg}").expect("valid template"));
        pb
    });
    let update_summary = |done: usize| {
        if let Some(pb) = &summary {
            pb.set_message(summary_message(done, hosts.len(), start));
            pb.tick();
        }
    };
    update_summary(0);

    let mut results: Vec<Option<ExecResult>> = hosts.iter().map(|_| None).collect();
    let mut remaining = hosts.len();
    while remaining > 0 {
        match rx.recv_timeout(TICK_INTERVAL) {
            Ok((i, result)) => {
                bars[i].finish_and_clear();
                let _ = mp.println(render_result_block(&result));
                results[i] = Some(result);
                remaining -= 1;
                update_summary(hosts.len() - remaining);
            }
            Err(RecvTimeoutError::Timeout) => {
                for (i, pb) in bars.iter().enumerate() {
                    if results[i].is_none() {
                        pb.set_message(running_message(&hosts[i], start));
                        pb.tick();
                    }
                }
                update_summary(hosts.len() - remaining);
            }
            Err(RecvTimeoutError::Disconnected) => break,
        }
    }
    if let Some(pb) = &summary {
        pb.finish_and_clear();
    }

    Ok(collect_results(results, hosts, command, start))
}

fn summary_message(done: usize, total: usize, start: Instant) -> String {
    styles::secondary()
        .apply_to(format!(
            "{done}/{total} done · {}",
            format_duration(start.elapsed())
        ))
        .to_string()
}

fn running_message(host: &str, start: Instant) -> String {
    format!(
        "{}  {}",
        styles::hostname().apply_to(host),
        styles::secondary().apply_to(format_duration(start.elapsed()))
    )
}

fn run_streaming_plain(hosts: &[String], command: &str) -> Result<Vec<ExecResult>> {
    let start = Instant::now();
    let rx = executor::run_parallel(hosts, command)?;
    let mut results: Vec<Option<ExecResult>> = hosts.iter().map(|_| None).collect();
    for _ in hosts {
        let Ok((i, result)) = rx.recv() else { break };
        println!("{}", render_result_block(&result));
        println!();
        results[i] = Some(result);
    }
    Ok(collect_results(results, hosts, command, start))
}

/// Unwraps per-host results, substituting an error result for any host whose
/// worker thread died without reporting (should not happen in practice).
pub fn collect_results(
    results: Vec<Option<ExecResult>>,
    hosts: &[String],
    command: &str,
    start: Instant,
) -> Vec<ExecResult> {
    results
        .into_iter()
        .enumerate()
        .map(|(i, result)| {
            result.unwrap_or_else(|| ExecResult {
                hostname: hosts[i].clone(),
                command: command.to_string(),
                stdout: String::new(),
                stderr: String::new(),
                err: Some(format!(
                    "error executing on {}: worker terminated",
                    hosts[i]
                )),
                duration: start.elapsed(),
            })
        })
        .collect()
}
